backend/analyze_tailcall_graph.ml

```diff
@@ -0,0 +1,416 @@
+    type label =
+      (* "Plain" edges are edges that are (very likely) not changing their
+         TCO behavior as a result of adding TCO inference. I.e., before TCO
+         inference, they were tail (nontail) if and only if after TCO
+         inference they are tail (nontail). *)
+      | Plain_tail_edge
+      | Plain_nontail_edge
+      (* "Inferred" edges are edges that might possibly change their TCO
+         behavior as a result of adding TCO inference. *)
+      | Inferred_tail_edge
+      | Inferred_nontail_edge
+
+    type t =
+      { from : Vertex.t
+      ; to_ : Vertex.t
+      ; label : label
+      }
+
+    module Tbl : Hashtbl.S with type key = t
+    module Hashset : Hashset.S with type elt = t
+  end
+
+  type t
+
+  val create : unit -> t
+  val reset : t -> unit
+  val successors : t -> Vertex.t -> Edge.Hashset.t
+  val find_or_add_vertex : t -> fn_name:string -> Vertex.t
+
+  val add_edge
+    :  t
+    -> from:Vertex.t
+    -> to_:Vertex.t
+    -> actual_position:[ `Tail | `Nontail ]
+    -> original_position:Typedtree.position_and_tail_attribute
+    -> unit
+
+  val print_dot : t -> Format.formatter -> unit
+end = struct
+  module Vertex = struct
+    module T = struct
+      type t =
+        | Unknown_fn
+        | Known_fn of
+            { id : int
+            ; name : string
+            }
+
+      let unknown = Unknown_fn
+
+      let is_unknown t =
+        match t with
+        | Unknown_fn -> true
+        | Known_fn _ -> false
+      ;;
+
+      let id t =
+        match t with
+        | Unknown_fn -> -1
+        | Known_fn { id; _ } -> id
+      ;;
+
+      let equal t1 t2 = id t1 = id t2
+      let hash t = Int.hash (id t)
+    end
+
+    include T
+
+    let to_dot_id t =
+      match t with
+      | Unknown_fn -> "unknown"
+      | Known_fn { id; _ } -> Int.to_string id
+    ;;
+
+    let to_dot_label t =
+      match t with
+      | Unknown_fn -> "<unknown>"
+      | Known_fn { name; _ } -> name
+    ;;
+
+    module Tbl = Hashtbl.Make (struct
+        include T
+      end)
+
+    module Hashset = Hashset.Make (Tbl)
+  end
+
+  module Edge = struct
+    module T = struct
+      type label =
+        | Plain_tail_edge
+        | Plain_nontail_edge
+        | Inferred_tail_edge
+        | Inferred_nontail_edge
+
+      type t =
+        { from : Vertex.t
+        ; to_ : Vertex.t
+        ; label : label
+        }
+
+      let label_equal l1 l2 =
+        match l1, l2 with
+        | Plain_tail_edge, Plain_tail_edge -> true
+        | Plain_nontail_edge, Plain_nontail_edge -> true
+        | Inferred_tail_edge, Inferred_tail_edge -> true
+        | Inferred_nontail_edge, Inferred_nontail_edge -> true
+        | _ -> false
+      ;;
+
+      let equal e1 e2 =
+        label_equal e1.label e2.label
+        && Vertex.equal e1.from e2.from
+        && Vertex.equal e1.to_ e2.to_
+      ;;
+
+      let hash e1 = Vertex.hash e1.from + Vertex.hash e1.to_
```

Comment:
Nit: is there a reason we exclude the label from the hash? It is possible to have two edges with the same source and destination, but different labels.

Also, if we end up making a `hash_label` function, we might want to pull the `label` type and its utilities to a submodule.

Reply:
I forgot that there could be multiple edges with different labels.

Added `Label` struct.